helix-term/src/commands/typed.rs

```diff
@@ -1410,47 +1410,51 @@ fn tutor(
+    if let Some(last_selection) = &cx.editor.last_selection {
+        let scrolloff = cx.editor.config().scrolloff;
+        let (view, doc) = current!(cx.editor);
+        doc.set_selection(view.id, last_selection.clone());
+        view.ensure_cursor_in_view(doc, scrolloff);
+        cx.editor.last_selection = None;
+    }
+}
+
 pub(super) fn goto_line_number(
     cx: &mut compositor::Context,
     args: &[Cow<str>],
     event: PromptEvent,
 ) -> anyhow::Result<()> {
     match event {
-        PromptEvent::Abort => {
-            if let Some(line_number) = cx.editor.last_line_number {
-                goto_line_impl(cx.editor, NonZeroUsize::new(line_number));
-                let (view, doc) = current!(cx.editor);
-                view.ensure_cursor_in_view(doc, line_number);
-                cx.editor.last_line_number = None;
-            }
-            return Ok(());
-        }
+        PromptEvent::Abort => abort_goto_line_number_preview(cx),
         PromptEvent::Validate => {
             ensure!(!args.is_empty(), "Line number required");
-            cx.editor.last_line_number = None;
+
+            if let Some(last_selection) = &cx.editor.last_selection {
+                let (view, doc) = current!(cx.editor);
+                view.jumps.push((doc.id(), last_selection.clone()));
+
+                cx.editor.last_selection = None;
+            }
         }
+        // When a user hits backspace and there are no numbers left,
+        // we can bring them back to their original selection. If they
+        // begin typing numbers again, we'll start a new preview session.
+        PromptEvent::Update if args.is_empty() => abort_goto_line_number_preview(cx),
         PromptEvent::Update => {
-            if args.is_empty() {
-                if let Some(line_number) = cx.editor.last_line_number {
-                    // When a user hits backspace and there are no numbers left,
-                    // we can bring them back to their original line
-                    goto_line_impl(cx.editor, NonZeroUsize::new(line_number));
-                    let (view, doc) = current!(cx.editor);
-                    view.ensure_cursor_in_view(doc, line_number);
-                    cx.editor.last_line_number = None;
-                }
-                return Ok(());
-            }
+            cx.editor.last_selection.get_or_insert_with(|| {
+                let (view, doc) = current!(cx.editor);
+                doc.selection(view.id).clone()
+            });
+
+            let scrolloff = cx.editor.config().scrolloff;
+            let line = args[0].parse::<usize>()?;
+            goto_line_without_jumplist(cx.editor, NonZeroUsize::new(line));
+
             let (view, doc) = current!(cx.editor);
-            let text = doc.text().slice(..);
-            let line = doc.selection(view.id).primary().cursor_line(text);
-            cx.editor.last_line_number.get_or_insert(line + 1);
+            view.ensure_cursor_in_view(doc, scrolloff);
```

Comment:
Stumbled on a neat "bug" in this last pass where we were passing `line_number` as the scrolloff value of `ensure_cursor_in_view`. Fixed in the last iteration.

Reply:
that could cause very weird issues. Nice catch